staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go

```diff
@@ -67,6 +73,15 @@ func (s *specAggregator) addLocalSpec(spec *spec.Swagger, localHandler http.Hand
 		handler:    localHandler,
 		spec:       spec,
 	}
+	// Record delegation target API services' names
+	if localHandler != nil {
+		s.delegationAPIServices = append(s.delegationAPIServices, name)
```

Comment:
can't this be derived from `s.openAPISpecs[name]`? Why an extra slice?

Reply:
We want to pass a list of names to the controller's [queue](https://github.com/kubernetes/kubernetes/blob/62060e1f046a888c475f656193219401fae35a0e/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go#L82), so that it can periodically sync openapi specs from delegation apiservers. ~~We didn't have a place that recorded the names of the delegation APIServices.~~

Actually, I thought we cannot use the keys (names) in `s.openAPISpecs`, because they consist of both delegation APIServices' names and aggregated APIServices' names, and there will be race condition if we pass all the keys to the queue. It turns out that we do the passing during controller initialization, so there should be no aggregated APIService's name.

I agree that the keys in `s.openAPISpecs` will be mostly the same as `s.delegationAPIServices ` during [controller initialization](https://github.com/kubernetes/kubernetes/blob/62060e1f046a888c475f656193219401fae35a0e/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go#L82) (except the [local spec with no handler](https://github.com/kubernetes/kubernetes/blob/f48e18faa4dc035cc927c6a2b34c83c8475b55fa/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go#L87-L88)-- it's a minor case, the controller will [drop the name](https://github.com/kubernetes/kubernetes/blob/f48e18faa4dc035cc927c6a2b34c83c8475b55fa/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go#L121-L123))